[IOTDB-2609] A new lossy encoding method based on frequency domain

543202718 opened the pull request:
This pull request adds a new lossy encoding method named FREQ. The key idea is that the time sequence is transformed to frequency domain and only the frequency components with high energy are reserved. The details of FREQ encoding are shown in https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=199536917

HTHou commented:
Hi, please create a Jira issue and link it in the title of this pr. Thanks so much!

HTHou commented on antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlLexer.g4:
```diff
@@ -664,6 +664,9 @@ TS_2DIFF
     : T S '_' '2' D I F F
     ;
 
+FREQ
+	: F R E Q
+	;
```
```suggestion
FREQ
    : F R E Q
    ;
```

543202718 replied:
Thanks

HTHou commented on tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  public void add(long x, int len) {
+    x = x & ~(ALL_MASK << len); // 保证x除最低的len位之外都是0
```
Use english?

543202718 replied:
Rewrite

HTHou commented on tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java:
```diff
@@ -0,0 +1,97 @@
+  public void add(long x, int len) {
+    x = x & ~(ALL_MASK << len); // 保证x除最低的len位之外都是0
+    while (len > 0) {
+      int m = len + cnt >= BITS_IN_A_BYTE ? BITS_IN_A_BYTE - cnt : len; // 向cache中插入的bit数
```
english?

HTHou commented on tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java:
```diff
@@ -0,0 +1,97 @@
+    cnt = 0;
+  }
+
+  /** 如果当前字节存在剩余位，则全部填充0 */
```
Same

543202718 replied:
Rewrite

HTHou commented on tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitReader.java:
```diff
@@ -0,0 +1,70 @@
+      if (bitCnt == BITS_IN_A_BYTE) {
+        next();
+      }
+      int m = len + bitCnt >= BITS_IN_A_BYTE ? BITS_IN_A_BYTE - bitCnt : len; // 从当前byte中读取的bit数
+      len -= m;
+      ret = ret << m;
+      byte y = (byte) (cache & MASKS[bitCnt]); // 运算时byte自动转化为int，需要&截取低位
+      y = (byte) ((y & 0xff) >>> (BITS_IN_A_BYTE - bitCnt - m)); // 逻辑右移，高位补0
```
Use english?

543202718 replied:
Rewrite

HTHou commented on server/src/assembly/resources/conf/iotdb-engine.properties:
```diff
@@ -730,6 +730,14 @@ timestamp_precision=ms
 # Datatype: long
 # slow_query_threshold=5000
 
+# Signal-noise-ratio (SNR) of FREQ encoding
+# Datatype: double
+# freq_snr=40.0
+
+# Block size of FREQ encoding
+# Datatype: integer
+# freq_block_size=1024
+
```
Add these parameters to `Reference/Config-Manual.md`.

543202718 replied:
Added

HTHou commented:
> The details of FREQ encoding are shown in http://ise.thss.tsinghua.edu.cn/confluence/pages/viewpage.action?pageId=64527651

Hi, can you please move this doc to https://cwiki.apache.org/confluence/display/IOTDB/Home? It will be convenient for more people to read.

543202718 commented:
> > The details of FREQ encoding are shown in http://ise.thss.tsinghua.edu.cn/confluence/pages/viewpage.action?pageId=64527651
> 
> Hi, would you like to move this doc to https://cwiki.apache.org/confluence/display/IOTDB/Home? It will be convenient for more people to read.

That's a good idea. But I don't have the right to edit on this website. Could you please move the doc or give me the right?

HTHou commented:
> > > The details of FREQ encoding are shown in http://ise.thss.tsinghua.edu.cn/confluence/pages/viewpage.action?pageId=64527651
> > 
> > 
> > Hi, would you like to move this doc to https://cwiki.apache.org/confluence/display/IOTDB/Home? It will be convenient for more people to read.
> 
> That's a good idea. But I don't have the right to edit on this website. Could you please move the doc or give me the right?

Sure. What is your apache confluence ID?

543202718 commented:
> > > > The details of FREQ encoding are shown in http://ise.thss.tsinghua.edu.cn/confluence/pages/viewpage.action?pageId=64527651
> > > 
> > > 
> > > Hi, would you like to move this doc to https://cwiki.apache.org/confluence/display/IOTDB/Home? It will be convenient for more people to read.
> > 
> > 
> > That's a good idea. But I don't have the right to edit on this website. Could you please move the doc or give me the right?
> 
> Sure. What is your apache confluence ID?

wanghy20

HTHou commented:
> > Sure. What is your apache confluence ID?
> 
> wanghy20

Done!

543202718 commented:
The details of FREQ encoding is now also shown in https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=199536917

HTHou reviewed (approved):
LGTM

HTHou commented on docs/UserGuide/Data-Concept/Encoding.md:
```diff
@@ -54,6 +54,12 @@ Usage restrictions: When using GORILLA to encode INT32 data, you need to ensure
 
+* FREQ
+
+FREQ encoding is lossy. It transforms the time sequence to the frequency domain and only reserve part of the frequency components with high energy. It is more suitable for sequence with obvious periodicity.
+
+> There are two parameters of FREQ encoding in the configuration file: `freq_snr` defines the signal-noise-ratio (SNR), `freq_block_size` defines the data size in a time-frequency transformation.
```
How do these two parameters affect the encoding performance and compression ratio?

543202718 replied:
The experimental results and analysis have been put on Apache confluence https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=199536917

HTHou replied:
Yes, I mean, to tell users how to use these parameters, we can show the conclusion in this user guide here. :)

HTHou replied:
Yes, I mean, to tell users how to use these parameters, we can show the conclusion in this user guide here. :)

543202718 replied:
The brief conclusion is added in the user guide.

HTHou reviewed (changes requested):
Hi, please fix the conflicts and consider add support for Python and C++ client.

See:
https://github.com/apache/iotdb/blob/master/client-py/iotdb/utils/IoTDBConstants.py
https://github.com/apache/iotdb/blob/master/client-cpp/src/main/Session.h

Also, please add tests in 
https://github.com/apache/iotdb/blob/master/integration/src/test/java/org/apache/iotdb/db/integration/IoTDBEncodingIT.java

543202718 commented:
> Hi, please fix the conflicts and consider add support for Python and C++ client.
> 
> See: https://github.com/apache/iotdb/blob/master/client-py/iotdb/utils/IoTDBConstants.py https://github.com/apache/iotdb/blob/master/client-cpp/src/main/Session.h
> 
> Also, please add tests in https://github.com/apache/iotdb/blob/master/integration/src/test/java/org/apache/iotdb/db/integration/IoTDBEncodingIT.java


Thanks. I have fixed the conflicts and added the support for Python and C++. Meanwhile, the tests are also added in IoTDBEncodingIT.

HTHou reviewed (approved):
LGTM
